pyro/distributions/omt_mvn.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+from torch.autograd import Function
+from torch.autograd.function import once_differentiable
+from torch.distributions import constraints
+
+from pyro.distributions.torch.multivariate_normal import MultivariateNormal
+
+
+class OMTMultivariateNormal(MultivariateNormal):
+    """Multivariate normal (Gaussian) distribution with OMT gradients w.r.t. both
+    parameters. Note the gradient computation w.r.t. the Cholesky factor has cost
+    O(D^3), although the resulting gradient variance is generally expected to be lower.
+
+    A distribution over vectors in which all the elements have a joint Gaussian
+    density.
+
+    :param torch.autograd.Variable loc: Mean.
+    :param torch.autograd.Variable scale_tril: Cholesky of Covariance matrix.
+    """
+    params = {"loc": constraints.real, "scale_tril": constraints.lower_triangular}
+    support = constraints.real
+    has_rsample = True
+    reparameterized = True
+
+    def __init__(self, loc, scale_tril):
+        assert(len(loc.size()) == 1), "OMTMultivariateNormal loc must be 1-dimensional"
+        assert(len(scale_tril.size()) == 2), "OMTMultivariateNormal scale_tril must be 2-dimensional"
+        covariance_matrix = torch.mm(scale_tril, scale_tril.t())
+        super(OMTMultivariateNormal, self).__init__(loc, covariance_matrix)
+        self.scale_tril = scale_tril
+
+    def rsample(self, sample_shape=torch.Size()):
+        return _OMTMVNSample.apply(self.loc, self.scale_tril, sample_shape + self.loc.shape)
+
+
+def sum_all_but_rightmost_dim(x):
```

Comment:
Consider adding a generalized version to pyro/distributions/util.py
```py
def sum_leftmost(x, dim):
    if dim < 0:
        dim = x.dim() - dim
    if dim == 0:
        return x
    return x.contiguous().view(-1, *x.shape[dim:]).sum(0)

x = torch.ones(2,3,4)
assert sum_leftmost(x, -2).shape == (3, 4)
assert sum_leftmost(x, -1).shape == (4,)
```

Reply:
Actually this would be useful for me as well :smile:

Reply:
done, thanks